### PR TITLE
[language][move] Added typing tests [part 1]

### PR DESCRIPTION
**Pull request by tnowacki**
## Motivation

- Added tests for typing pass.
- Tried to cover +/- cases for each check. Covers:
  - Constraint solving/instantiation on user defined types
  - Binds/Decls/Assigns
  - Returns
  - Scoping/Shadowing
  - Simple values

## Test Plan

ran em

### CODE REVIEW

File: language/move-lang/src/typing/translate.rs [hunk 1]
```diff
@@ -916,6 +924,13 @@ fn close_scope(
     sp!(_, bs_): &mut T::BindList,
 ) {
     bs_.iter_mut().for_each(|b| close_scope_bind(context, b));
+
+    // remove new locals from inner scope
+    for (new_local, _) in declared.iter().filter(|(v, _)| !old_locals.contains_key(v)) {
+        context.locals.remove(&new_local);
+    }
```
Comment (tzakian) [inline language/move-lang/src/typing/translate.rs]: Might be more idiomatic here to write this as a `for_each`?

File: language/move-lang/tests/move_check/typing/assign_unpack_references.move [hunk 2]
```diff
@@ -0,0 +1,28 @@
+module M {
+    struct S { f: u64 }
+    struct R { s1: S, s2: S }
+
+    t0() {
+        let f;
+        let s2;
+        R { s1: S { f }, s2 } = R { s1: S{f: 0}, s2: S{f: 1} };
+        f = 0;
+        s2 = S { f: 0 }
```
Comment (tzakian) [inline language/move-lang/tests/move_check/typing/assign_unpack_references.move]: Do you have a similar test to this but where you annotate with the wrong types in the declaration?

File: language/move-lang/tests/move_check/typing/bind_duplicate_binding.move [hunk 3]
```diff
@@ -0,0 +1,8 @@
+module M {
```
Comment (tzakian) [inline language/move-lang/tests/move_check/typing/bind_duplicate_binding.move]: Why is this a typing test? Shouldn't this be caught in expansion/naming?

File: language/move-lang/tests/move_check/typing/bind_wrong_arity.exp [hunk 4]
```diff
@@ -0,0 +1,44 @@
+error: 
+
+   ┌── tests/move_check/typing/bind_wrong_arity.move:5:13 ───
+   │
+ 5 │         let x: () = ();
+   │             ^ Invalid list-binding. The list-binding expected 1 argument(s) but got 0
```
Comment (tzakian) [inline language/move-lang/tests/move_check/typing/bind_wrong_arity.exp]: List binding is confusing to me. Is there a better name that could be chosen for this?

File: language/move-lang/tests/move_check/typing/return_type_last_exp_invalid.move [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+    }
+
+    t2(): (u64, bool) {
+        (0, false, R{})
+    }
+
+    t3(): (u64, bool, R, bool) {
+        (0, false, R{})
+    }
+}
```
Comment (tzakian) [inline language/move-lang/tests/move_check/typing/return_type_last_exp_invalid.move]: Might also be worthwhile to have a test for same number of values but in the incorrect order to the type (e.g. return `(0, false, R)` to a return type of `(bool, u64, R)`).

File: language/move-lang/tests/move_check/typing/bind_duplicate_binding.move (hunk 3 above)
Comment (tnowacki) [reply language/move-lang/tests/move_check/typing/bind_duplicate_binding.move]: Its caught in typing currently ¯\_(ツ)_/¯

File: language/move-lang/tests/move_check/typing/assign_unpack_references.move (hunk 2 above)
Comment (tnowacki) [reply language/move-lang/tests/move_check/typing/assign_unpack_references.move]: ... maybe I'll double check

File: language/move-lang/src/typing/translate.rs (hunk 1 above)
Comment (tnowacki) [reply language/move-lang/src/typing/translate.rs]: It takes up more lines with for_each ¯\_(ツ)_/¯

File: language/move-lang/tests/move_check/typing/bind_wrong_arity.exp (hunk 4 above)
Comment (tnowacki) [reply language/move-lang/tests/move_check/typing/bind_wrong_arity.exp]: I don't use the term tuple anywhere so I was kind of stuck. I refer to them as expression-lists on the expression errors

**Pull request closed without merging**